### PR TITLE
[docs] CUDA generator output depends on the device's SM count; note it in the reproducibility note and the seeding docstrings

### LINKED ISSUE #196613: [docs] CUDA generator output depends on the device's SM count: seeding does not make device-side sampling reproducible across GPU models or MIG slices
### 📚 The doc issue

`torch.rand`/`torch.randn` (and every distribution kernel built on `distribution_nullary_kernel`) with a **seeded CUDA generator** produce different values on GPUs with a different number of SMs (or a different `maxThreadsPerMultiProcessor`) as soon as the tensor exceeds one wave of the launch grid, i.e. when

```
numel > 256 * multiProcessorCount * (maxThreadsPerMultiProcessor / 256)
```

This is by design and has been true for a long time — the RNG (Philox) is deterministic; what depends on the device is the mapping of output elements to Philox counters, because the grid is capped by the SM count:

```cpp
// aten/src/ATen/native/cuda/DistributionTemplates.h (v2.13.0), calc_execution_policy, lines 49-61
dim3 grid((numel + block_size - 1) / block_size);
uint32_t blocks_per_sm = at::cuda::getCurrentDeviceProperties()->maxThreadsPerMultiProcessor / block_size;
grid.x = std::min(static_cast<uint32_t>(at::cuda::getCurrentDeviceProperties()->multiProcessorCount) * blocks_per_sm, grid.x);
// lines 65-88: thread idx seeds curand_init(seed, idx, offset); element
// li = linear_index + blockDim.x*gridDim.x*ii takes component ii of that thread's k-th curand_uniform4()
```

But it is not stated where users look — the [Reproducibility note](https://pytorch.org/docs/stable/notes/randomness.html), `torch.Generator`, `torch.manual_seed`, `torch.cuda.manual_seed` — and the note's advice to seed the generator reads as if seeded device-side sampling were portable across GPUs. It is not: the same seed gives a different initialisation, a different dropout mask, a different augmentation sample on another SKU, on a MIG slice of the same card, or after a hardware refresh.

### What I measured (torch 2.13.0+cu129, `torch.Generator(device="cuda").manual_seed(0)`, `torch.rand(n)`, sha256 of the raw fp32 bytes)

| card (SMs × blocks/SM) | grid cap | `rand(2^16)` | `rand(2^20)` |
|---|---|---|---|
| L40 (142 × 6) | 852 | `2ccd6bcb5339…` | `28a2fbc8ed96…` |
| L40S (142 × 6) | 852 | `2ccd6bcb5339…` | `28a2fbc8ed96…` |
| H100 PCIe (114 × 8) | 912 | `2ccd6bcb5339…` | `ebd21badc8cf…` |
| A10G (80 × 6) | 480 | `2ccd6bcb5339…` | `ec3fe0ff87f0…` |

Below one wave (2^16 = 256 blocks) every card agrees; at 2^18 and above the L40, the H100 and the A10G all disagree, and so do the 1g.10gb / 3g.40gb MIG slices of one H100 against the whole card. A CPU re-implementation of Philox4x32-10 + cuRAND's `curand_init(seed, subsequence=idx, offset)` layout + `_curand_uniform` (integer → float first, then `x * 2^-32 + 2^-33`) + the grid formula above reproduces each of the three cards' output **bit for bit at every size tested (2^12 … 2^22)** when given that card's `(multiProcessorCount, maxThreadsPerMultiProcessor)`, and produces the other card's bytes when given the other card's numbers — so the dependence is exactly the launch geometry and nothing else. (`randn` shows the same card-dependence; I did not reproduce its Box–Muller bits.)

### Suggested change

A sentence in the Reproducibility note (and in the `torch.Generator` / `torch.cuda.manual_seed` docstrings), roughly:

> Values drawn from a CUDA generator depend on the device's multiprocessor count (and its maximum threads per multiprocessor) once a tensor is larger than one wave of the launch grid. Seeding does **not** make device-side sampling reproducible across different GPU models or MIG configurations. For cross-device reproducibility, sample on the CPU and copy to the device, or record the device properties alongside the seed.

I am not proposing to change the kernel — that would change every existing user's streams — only to say what it does.

### Versions

torch 2.13.0+cu129 (also observed with the cu126 build), driver 580.126.09 / 570.x; NVIDIA L40, L40S, H100 PCIe (and MIG slices), A100 SXM4, A10, A10G, L4, T4 — seven distinct (architecture, SM count) classes gave seven distinct `randn(2^20)` digests and the same digest whenever the SM count was the same.

<details><summary>CPU reproduction (numpy), ~60 lines</summary>

```python
import hashlib, numpy as np
M0, M1, W0, W1 = np.uint32(0xD2511F53), np.uint32(0xCD9E8D57), 0x9E3779B9, 0xBB67AE85
def mulhilo(a, b):
    p = np.uint64(a) * b.astype(np.uint64); return (p >> np.uint64(32)).astype(np.uint32), (p & np.uint64(0xFFFFFFFF)).astype(np.uint32)
def philox4x32_10(c0, c1, c2, c3, k0, k1):
    for _ in range(10):
        hi0, lo0 = mulhilo(M0, c0); hi1, lo1 = mulhilo(M1, c2)
        c0, c1, c2, c3 = hi1 ^ c1 ^ np.uint32(k0), lo1, hi0 ^ c3 ^ np.uint32(k1), lo0
        k0, k1 = (k0 + W0) & 0xFFFFFFFF, (k1 + W1) & 0xFFFFFFFF
    return c0, c1, c2, c3
def torch_cuda_rand(n, sm_count, max_threads_per_sm, seed=0):
    grid = min(sm_count * (max_threads_per_sm // 256), (n + 255) // 256); stride = 256 * grid
    n_iter = (n - 1) // (stride * 4) + 1
    idx = np.arange(stride, dtype=np.uint64); out = np.zeros(n_iter * stride * 4, dtype=np.float32)
    for k in range(n_iter):                       # curand_init(seed, idx, 0): ctr = (k, 0, idx_lo, idx_hi), key = seed
        r = philox4x32_10(np.full(stride, k, np.uint32), np.zeros(stride, np.uint32),
                          (idx & np.uint64(0xFFFFFFFF)).astype(np.uint32), (idx >> np.uint64(32)).astype(np.uint32),
                          seed & 0xFFFFFFFF, seed >> 32)
        for ii in range(4):                       # _curand_uniform: uint -> float first, then x*2^-32 + 2^-33
            xf = r[ii].astype(np.float32)
            v = (xf.astype(np.float64) * 2.0**-32 + 2.0**-33).astype(np.float32)
            v = np.where(v == np.float32(1.0), np.float32(0.0), v)      # uniform_kernel's reverse_bound_value
            out[k * stride * 4 + ii * stride + np.arange(stride)] = v
    return out[:n]
# torch side, on the GPU:  g = torch.Generator(device="cuda").manual_seed(0); t = torch.rand(1 << 20, device="cuda", generator=g)
#                          hashlib.sha256(t.cpu().numpy().tobytes()).hexdigest()
print(hashlib.sha256(torch_cuda_rand(1 << 20, 142, 1536).tobytes()).hexdigest()[:12])   # L40:  28a2fbc8ed96
print(hashlib.sha256(torch_cuda_rand(1 << 20, 114, 2048).tobytes()).hexdigest()[:12])   # H100: ebd21badc8cf
print(hashlib.sha256(torch_cuda_rand(1 << 20, 80, 1536).tobytes()).hexdigest()[:12])    # A10G: ec3fe0ff87f0
```
</details>

cc @svekars @sekyondaMeta @AlannaBurke @pbelevich

### PR DESCRIPTION
**Pull request by iamgahr**
Documentation only. Fixes #196613.

Seeding a CUDA generator makes device-side sampling reproducible on the same GPU model, not across GPU models — and the documentation currently lets a reader believe otherwise (the reproducibility note's advice to seed, `torch.cuda.manual_seed`, `torch.Generator(device='cuda')`).

**Mechanism** (`aten/src/ATen/native/cuda/DistributionTemplates.h`, `calc_execution_policy` and the sampling kernels): the Philox subsequence is the *thread* index, and the launch grid is `min(ceil(numel / 256), multiProcessorCount * (maxThreadsPerMultiProcessor / 256))` blocks. Below one wave of threads every card computes the same (counter, key) for every element; above it the mapping follows the grid, which follows the SM count.

**Measured** (torch 2.13.0+cu129): a CPU implementation of that mapping (Philox4x32-10, cuRAND's `curand_init(seed, subsequence, offset)` layout, `_curand_uniform`) reproduces `torch.rand(n, device="cuda", generator=Generator(device="cuda").manual_seed(0))` bit for bit at n = 2^12 … 2^22 on an A10G (80 SMs × 6 blocks/SM), an L40 and an L40S (142 × 6) and an H100 PCIe (114 × 8), given only each card's SM count and threads per SM; across an eleven-instance training campaign, `randn(2^20)` from a seeded device generator gave seven distinct outputs for seven distinct (architecture, SM count) classes and identical outputs for the same SM count; MIG slices of one H100 (14 and 46 SMs) differ from the whole card. #194676's empirically found T4/L4 boundary of 40,960 elements is exactly 256 × 40 × 4 by this formula; #84234 and #87992 (2022) are the same mechanism. Details and the script are in #196613.

**Change (trimmed after review):** three sentences in `docs/source/notes/randomness.md` (under "PyTorch random number generator") stating that seeding makes device-side sampling reproducible on the same GPU, that the same seed is not guaranteed to match on a different GPU model or a MIG partition and PyTorch does not test for it, and that a tensor which must be identical across devices should be sampled on the CPU with a seeded `torch.Generator` and copied; plus a one-sentence `.. note::` on `torch.cuda.manual_seed` pointing to it. No mechanism, no numbers, no guarantee claimed; the measurement and the counter-layout derivation stay in #196613.

cc @kurtamohler (reviewed #194676), @svekars (docs)

### CODE REVIEW

File: docs/source/notes/randomness.md [hunk 1]
```diff
@@ -44,6 +44,17 @@ It is also possible to obtain identical results from an operation that uses
...
+Seeding a CUDA generator makes device-side sampling reproducible on the same GPU
+model, not across GPU models. For tensors larger than one launch wave, that is
+more than `256 * multiProcessorCount * (maxThreadsPerMultiProcessor // 256)`
+elements (for example 40,960 on a T4, 218,112 on an L40, 233,472 on an H100 PCIe),
+the mapping of tensor elements to Philox counters follows the launch grid, which
+is capped by the device's number of SMs. The same seed therefore produces
+different values on GPUs with a different number of SMs, including MIG partitions
+of one card, while tensors below that size are identical on every GPU. To
+reproduce sampled tensors across devices, sample on the CPU and copy the result
+to the device.
```
Comment (kurtamohler) [inline docs/source/notes/randomness.md]: In my opinion this contains too much unnecessary implementation detail.

More importantly I'm not sure that we can safely guarantee that random tensors below a certain size are the same across every GPU, or that every CPU will generate the same results. You could be right that the same results will be generated in these cases, but there's not really any way we can test this in CI, so we wouldn't have anything in place to catch if some change ever breaks this guarantee.

Furthermore, I question the value of this guarantee given that different GPUs are certainly give different results for some operations in pytorch. I'm not sure if there's much utility in guaranteeing that random generators will give the same results when other parts of a user's application won't.

One thing we could maybe do is mention in the docs that if you want to get identical random generation on different GPUs, here's what you can try, but you'll have to check it yourself because we don't actually guarantee that it will be the same.

**Review by kurtamohler — CHANGES_REQUESTED** [review]

File: docs/source/notes/randomness.md (hunk 1 above)
Comment (iamgahr) [reply docs/source/notes/randomness.md]: Trimmed as suggested, thanks. The note now states the non-guarantee and the workaround only; the formula, the numbers and the mechanism are gone from the docs (they stay in #196613 for anyone who wants them), and the `Generator` and `manual_seed_all` notes are dropped.

Two data points for your judgment, not for the docs:

1. Below one launch wave the outputs were identical on every GPU we tried (A10G, L40, L40S, H100 PCIe, and MIG slices of an A100/H100), and a CPU reimplementation of the counter layout reproduces each card's bytes from its own SM count. So the mapping *is* testable on a single GPU in CI, against that CPU reference, if a guarantee were ever wanted. I agree it should not be promised until something like that exists.

2. On value: once cuBLASLt algorithm selection is pinned, this mapping was the only remaining cross-GPU difference in our fp32 training runs (Ampere, Ada, Hopper, 300 steps). Users doing cross-device replay hit exactly this, which is why the one-sentence workaround seemed worth keeping.

**Review by ngimel — APPROVED** [review]

**Review by ngimel — APPROVED** [review]
